### ISSUE TITLE
Why inplace operation in `AdaptiveSoftmax`

### ISSUE THREAD

**Issue by Stonesjtu** [pytorch/fairseq#760]
I used `log_output` for training purpose, but pytorch complains about inplace operation on `backward`.

For lines 204 and 205:
https://github.com/pytorch/fairseq/blob/e6422528dae0b899848469efe2dc404c1e639ce9/fairseq/modules/adaptive_softmax.py#L200-L208

1. why doing a `copy_` on tailout
2. why inplace add_

I changed to 
```python
tail_output = ...
... = self.lsm(tail_out) + tail_priors[idxs, i, None]
```
And it works for training.

**Comment by myleott** [comment]
I suspect it's to save memory, but cc @alexeib to confirm.

One option is to add an inplace flag to get_log_prob, so that people can turn it on/off as needed.

**Comment by alexeib** [comment]
yeah it is to save memory. have you noticed any difference in GPU memory usage or the max possible batch size after changing it to non inplace ops?

**Comment by Stonesjtu** [comment]
1. The current implementation is not working in `training` mode.
2. In evaluation mode, if we set the `torch.no_grad` env, the memory overhead should equal to the `tail_priors[idxs, i ,None]` (temporarily for the out-of-place `add`).

Actually `add` operation does not require saving operands for future BP, but pytorch does raise an error for any inplace operators.

**Comment by myleott** [comment]
Thanks, a PR with some comparison of memory usage would be great!
